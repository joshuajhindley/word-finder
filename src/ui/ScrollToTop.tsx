import React, { useEffect, useState } from 'react'
import { BiArrowFromBottom } from 'react-icons/bi'
import classNames from 'classnames'

const scrollToTop = () => {
  window.scrollTo({ top: 0, behavior: 'smooth' })
}

class ScrollIcon extends React.Component<any, any> {
  render() {
    return (
      <div className={this.props.className} onClick={scrollToTop}>
        <BiArrowFromBottom />
      </div>
    )
  }
}

export default function ScrollToTop(props: any) {
  const { darkMode } = props
  const [isVisible, setVisisble] = useState(false)

  const toggleVisibility = () => {
    if (window.scrollY > 100) {
      setVisisble(true)
    } else {
      setVisisble(false)
    }
  }

  useEffect(() => {
    window.addEventListener('scroll', toggleVisibility)
    return () => window.removeEventListener('scroll', toggleVisibility)
  }, [])

  const className = classNames('scroll-to-top', { dark: darkMode, visible: isVisible })
  const iconProps = { toggleVisibility, className }

  return <ScrollIcon {...iconProps} />
}
