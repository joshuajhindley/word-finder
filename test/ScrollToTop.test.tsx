import { mount } from 'enzyme'
import { act } from 'react-dom/test-utils'
import ScrollToTop from '../src/ui/ScrollToTop'

describe('tests the ScrollToTop component', () => {
  Object.defineProperty(window, 'scrollTo', {
    writable: true,
    value: jest.fn().mockImplementation(values => {
      global.scrollY = values.top
    })
  })

  it('matches the snapshot', () => {
    const wrapper = mount(<ScrollToTop />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders without crashing', () => {
    const wrapper = mount(<ScrollToTop />)
    expect(wrapper.find('.scroll-to-top').length).toBe(2)
  })

  it('does not render the scroll to top button when the window is at or near the top', async () => {
    const wrapper = mount(<ScrollToTop />)
    expect(wrapper.find('.scroll-to-top').length).toBe(2)
    expect(wrapper.find('.visible').length).toBe(0)
    global.scrollTo({ top: 50 })
    expect(global.scrollY).toEqual(50)

    const toggleVisibility = wrapper.find('ScrollIcon').prop('toggleVisibility') as Function
    await act(async () => {
      await new Promise(resolve => setTimeout(resolve))
      toggleVisibility()
      wrapper.update()
    })

    expect(wrapper.find('.scroll-to-top').length).toBe(2)
    expect(wrapper.find('.visible').length).toBe(0)
  })

  it('renders the scroll to top button when the window is not near the top', async () => {
    const wrapper = mount(<ScrollToTop />)
    global.scrollTo({ top: 200 })
    expect(global.scrollY).toEqual(200)

    const toggleVisibility = wrapper.find('ScrollIcon').prop('toggleVisibility') as Function
    await act(async () => {
      await new Promise(resolve => setTimeout(resolve))
      toggleVisibility()
      wrapper.update()
    })

    expect(wrapper.find('.scroll-to-top').length).toBe(2)
    expect(wrapper.find('.visible').length).toBe(2)
  })

  it('scrolls to the top when the scroll to top button is clicked', async () => {
    const wrapper = mount(<ScrollToTop />)
    global.scrollTo({ top: 200 })
    expect(global.scrollY).toEqual(200)

    wrapper.find('.scroll-to-top').at(1).simulate('click')
    const toggleVisibility = wrapper.find('ScrollIcon').prop('toggleVisibility') as Function
    await act(async () => {
      await new Promise(resolve => setTimeout(resolve))
      toggleVisibility()
      wrapper.update()
    })

    expect(global.scrollY).toEqual(0)
    expect(wrapper.find('.visible').length).toBe(0)
  })
})
