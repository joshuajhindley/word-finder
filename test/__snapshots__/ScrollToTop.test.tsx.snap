// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests the ScrollToTop component matches the snapshot 1`] = `
<ScrollToTop>
  <ScrollIcon
    className="scroll-to-top"
    toggleVisibility={[Function]}
  >
    <div
      className="scroll-to-top"
      onClick={[Function]}
    >
      <BiArrowFromBottom>
        <IconBase
          attr={
            Object {
              "viewBox": "0 0 24 24",
            }
          }
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18h12v2H6zm6-14.414-6.707 6.707 1.414 1.414L11 7.414V16h2V7.414l4.293 4.293 1.414-1.414z"
              key="0"
            />
          </svg>
        </IconBase>
      </BiArrowFromBottom>
    </div>
  </ScrollIcon>
</ScrollToTop>
`;
